Refactor to move positionChanged signal to Timeline

regulus79 opened the pull request:
Previously, this PR simply added a new signal to the `Timeline` class and had it emitted by the `TimeLineWidget` class in order to solve #7351.

However, as messmerd pointed out in his review comments, that was quite a hacky solution, and ideally the `positionChanged` signal would be solely managed by the backend `Timeline` class, while the frontend `TimeLineWidget` class would connect to those signals, instead of the other way around.

This PR is no longer a simple bugfix, but instead a refactoring of the `Timeline`/`TimeLineWidget` signal/slots.

## Changes

- NEW: Removed `PlayPos`, and instead store the timeline position in a `TimePos` along with a separate frame offset counter variable. The functions to set the ticks/timepos in `Timeline` emit the `positionChanged` signal. (Also, may emit `positionJumped` signal if the ticks were forcefully set--see below)
- The `positionChanged` signal and `updatePosition` slot were removed from `TimeLineWidget` and moved to `Timeline`.
- The connections to `Timeline::positionChanged` were made into `Qt::QueuedConnections` to prevent errors when the audio thread changes the `PlayPos` which would immediately cause gui code to run if it were a direct connection.
- The `pos()` method and `PlayPos m_pos` were removed from `TimeLineWidget`;
- The constructor for `TimeLineWidget` no longer requires a `PlayPos` reference.
- Since each `TimeLineWidget` stores a reference to its `Timeline`, a new method was added, `model()`, for other classes to access it.
- Removed array of `PlayPos`es in `Song`. Now each `Timeline` holds their respective ~~`PlayPos`~~`TimePos`.
- `Song`'s methods for `getPlayPos` were changed to return the ~~`PlayPos`~~`TimePos` of the respective `Timeline`. The non-const versions of the methods were removed because `Timeline` does not expose its `TimePos` to write.
- All of the places where `Timeline`s are used were updated, along with their calls to access/modify the `PlayPos`. For example, occurrences of `m_timeline->pos()` were replaced with `m_timeline->model()->getPlayPos()`, and calls to `m_timeline->pos().setTicks(ticks)` were changed to `m_timeline->model()->setTicks(ticks)`.
- ALSO: Removed `m_elapsedMilliseconds`, `m_elapsedBars`, and `m_elapsedTicks` from `Song`. The elapsed milliseconds is now handled individually by each `Timeline`.
- NEW: The `m_jumped` variable has been removed from `Timeline`. Now jumped events emit `Timeline::positionJumped` automatically whenever the ticks are forcefully set. This means it is no longer necessary to call `timeline->setJumped(true)` or `timeline->setFrameOffset(0)` whenever the playhead position is forcefully set. Many places in the codebase were already missing these calls, so this may fix some unknown bugs.


Minor note: The members `m_timeline` or `m_timeLine` in places like `PianoRoll` or `SongEditor` should honestly really be renamed something like `m_timeLineWidget`, as they are not `Timeline` objects.

<details>
<summary>Old PR Description</summary>
This pull request fixes #7351 by changing the PianoRoll's connection from `m_timeline` to `&Engine::getSong()->getTimeline(Song::PlayMode::Song)`, and additionally adding the signal `positionChanged` to the `Timeline` class. 

The new signal is necessary because `getTimeline()` returns a `Timeline`, which does not have a `positionChanged` signal, while `m_timeline` is a `TimeLineWidget` which does.

~~This pr is marked as a draft, since after fixing the bug I found another one: When pressing stop after record-play, the piano roll play head does not go back to the start, but instead stops at its current position.~~ Edit: I have decided to leave fixing that bug to a future pr, as it appears that it existed prior to the regression. Instead, this pr simply fixes the pianoroll playhead movement.
</detals>

messmerd commented on src/gui/editors/PianoRoll.cpp:
```diff
@@ -290,7 +290,7 @@ PianoRoll::PianoRoll() :
 
 	// update timeline when in record-accompany mode
-	connect(m_timeLine, &TimeLineWidget::positionChanged, this, &PianoRoll::updatePositionAccompany);
+	connect(&Engine::getSong()->getTimeline(Song::PlayMode::Song), SIGNAL(positionChanged(const lmms::TimePos&)), this, SLOT(updatePositionAccompany(const lmms::TimePos&)));
```
```suggestion
	connect(&Engine::getSong()->getTimeline(Song::PlayMode::Song), &Timeline::positionChanged, this, &PianoRoll::updatePositionAccompany);
```
We're trying to avoid using the `SIGNAL` and `SLOT` macros in new code

messmerd replied:
Actually, there may be a better approach than adding a signal to `Timeline`... Let me think this over some more

messmerd replied:
I don't really like how `Timeline` and `TimeLineWidget` have implemented their signals and slots for changes in position. I think ideally `Timeline` should be the model and only it should emit any `positionChanged` signals, while `TimeLineWidget` should be the view and its `updatePosition` slot should not emit any signals. The way it is currently, the `TimeLineWidget` is trying to do the job of both a model and a view, and that complicates things.

messmerd replied:
@DomClark Does that sound right to you? It looks to me like a proper solution to this bug may require rethinking how the timeline position signals/slots are implemented.

regulus79 replied:
I have now updated to PR to refactor everything in order to move the `positionChanged` signal from `TimeLineWidget` to `Timeline`.

Rossmaxx commented:
Fix looks simple. @messmerd did this fix your issue?

messmerd commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -273,23 +272,27 @@ void AutomationEditor::keyPressEvent(QKeyEvent * ke )
+			{
+				m_timeLine->timeline()->setTicks(0);
+			}
+			m_timeLine->timeline()->updatePosition();
```
It looks like `setTicks` already does the job of this method, so this would be emitting the same signal twice.

messmerd commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -273,23 +272,27 @@ void AutomationEditor::keyPressEvent(QKeyEvent * ke )
-			m_timeLine->pos() += 16;
-			m_timeLine->updatePosition();
+			m_timeLine->timeline()->setTicks(m_timeLine->timeline()->getTicks() + 16);
+			m_timeLine->timeline()->updatePosition();
```
Same here

messmerd commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -273,23 +272,27 @@ void AutomationEditor::keyPressEvent(QKeyEvent * ke )
-			m_timeLine->pos().setTicks( 0 );
-			m_timeLine->updatePosition();
+			m_timeLine->timeline()->setTicks(0);
+			m_timeLine->timeline()->updatePosition();
```
And here

messmerd reviewed (commented):
When I ran it, I saw this output in the console:
```
QObject::connect: Cannot queue arguments of type 'lmms::TimePos'
(Make sure 'lmms::TimePos' is registered using qRegisterMetaType().)
```

regulus79 commented on src/core/Song.cpp:
```diff
@@ -121,6 +121,7 @@ Song::Song() :
 
 	for (auto& scale : m_scales) {scale = std::make_shared<Scale>();}
 	for (auto& keymap : m_keymaps) {keymap = std::make_shared<Keymap>();}
+	for (auto& timeline : m_timelines) { connect(this, &Song::bufferProcessed, &timeline, &Timeline::updatePosition, Qt::QueuedConnection); }
```
I agree, I think the `bufferProcessed` signal and `updatePosition` slot can probably be removed.

Except that, when I remove them and instead rely on `setTicks(ticks)` to emit the signal, the `processNextBuffer` function in `Song` does not update the playhead position, so it stays still during playback. I believe this may be due to `processNextBuffer` being on a different thread?

But then it somehow fixes itself when I specify `Qt::DirectConnection` in all the `connect()`s. I'm not sure why.

regulus79 commented on src/core/Song.cpp:
```diff
@@ -186,6 +187,7 @@ void Song::setTimeSignature()
 
 void Song::savePlayStartPosition()
 {
+	// TODO simplify
```
I originally thought that it was repetitive to call a function on the `Timeline` while passing in the `Timeline`'s `PlayPos`, but now I realize that there in another instance of `setPlayStartPosition` in lmms which passes in a constant, so it wouldn't be possible to simplify this function to just set the play start position to whatever the current PlayPos is. In any case, I have now removed the TODO comment.

regulus79 commented:
The `QObject::connect: Cannot queue arguments of type 'lmms::TimePos'` warnings appear to have gone away after I changed the connections to `Qt::DirectConnection`. However, I do not know if this will impact the performance of the audio thread, as it may cause the gui to be updated and interrupt the buffer processing(?) However, using `Qt::QueuedConnection` seems to fail to actually recieve the signals, and I'm not sure why.

regulus79 commented:
I found a solution to the warnings by simply removing the parameter from the signal and instead having each editor query the current position of the timeline themselves.

sakertooth commented:
Here's a sketch of how I think this functionality should work (for my own understanding and possibly anybody elses). I'll try to complete the review at some point.

![image](https://github.com/user-attachments/assets/03229047-cbc8-4f20-91d1-d172eae36f7e)

sakertooth commented:
> Here's a sketch of how I think this functionality should work (for my own understanding and possibly anybody elses). I'll try to complete the review at some point.

Possible revision. Just using `std::atomic`?

regulus79 commented:
Okay, this is difficult.

**Originally**, I had it so that `Timeline` emitted `Timeline::positionChanged` whenever `Timeline::setPlayPos` or `Timeline::setTicks` was called. `Timeline::positionChanged` was then connected to all the editors with a `Qt::QueuedConnection`. That worked well, but sakertooth experienced segmentation faults which I could not reproduce. Additionally, the audio thread causing the `Timeline` to send out signals to the gui thread whenever a buffer was processed didn't seem like a great idea. They were all queued connections, so theoretically it should be fine, but I don't know.

So then sakertooth had the idea of routing all the connections to the gui editors through `TimeLineWidget`, which would be passing along the signals from `Timeline`. This could have been an improvement by reducing the number of `Qt::QueuedConnection`s, but it does not solve the issue of emitting signals across threads (which...may have been the cause of the segfaults? I do not know, I cannot reproduce them).

Another potential solution which sakertooth suggested is using `std::atomic`. This is a good idea to prevent race conditions, such as if the song is playing while the user drags on the timeline. However, removing all the positionChanged signals, while it simplifies the threading, means that we would have to reintroduce the `QTimer` to the `TimeLineWidget`s in order to update them. This is suboptimal. It is possible, and if it solves the segfaults, maybe it's the way to go. But it's not good.

What should be done?

sakertooth commented:
> So then sakertooth had the idea of routing all the connections to the gui editors through TimeLineWidget, which would be passing along the signals from Timeline. This could have been an improvement by reducing the number of Qt::QueuedConnections, but it does not solve the issue of emitting signals across threads (which...may have been the cause of the segfaults? I do not know, I cannot reproduce them).

Coming up with sound designs is probably my biggest weakness. The idea I mentioned could work, but as more time passes I continue to fail to see a critical issue using queued connections that emit the `positionChanged` signal directly from `Timeline`. That being said, this is what we are currently doing already I believe, so I'll just work on completing my review since the refactor is working in a way that makes the most sense already.

sakertooth commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -134,13 +134,10 @@ AutomationEditor::AutomationEditor() :
-	connect(this, &AutomationEditor::positionChanged, m_timeLine, &TimeLineWidget::updatePosition);
-	connect( m_timeLine, SIGNAL( positionChanged( const lmms::TimePos& ) ),
-			this, SLOT( updatePosition( const lmms::TimePos& ) ) );
+	connect(m_timeLine->model(), &Timeline::positionChanged, this, &AutomationEditor::updatePosition, Qt::QueuedConnection);
```
I don't know how safe it is to remove stuff like `connect(this, &AutomationEditor::positionChanged, m_timeLine, &TimeLineWidget::updatePosition);` Anything that is still emitting the `positionChanged` signal in `AutomationEditor` (like `AutomationEditor::horScrolled`) will be broken. This is a similar situation I have encountered in your PR with classes like `PianoRoll`, `SongEditor`, etc. It looks like these two signals are separate, so maybe you should keep that connection instead of removing it?

regulus79 replied:
Whooops, I don't think I meant to remove that first `connect`.

regulus79 commented on include/PlayPos.h:
```diff
@@ -0,0 +1,54 @@
+/*
+ * PlayPos.h - declaration of class PlayPos, a TimePos with additional features useful for Timelines
+ *
```
I'm not sure `TimePos` can easily be refactored to store frames as its smallest unit of time, unless we allow for fractional frames. If the number of frames in a bar is not a whole number, then it will have to be rounded. That's fine, but if the bpm is changed, then every single TimePos will have to be updated at once so that the number of frames matches the correct tick amount. I feel like a better long-term solution would be to allow float tick values, but that's just me.

But I agree, the jumped variable can probably be combined with the `positionChanged` signal.

sakertooth replied:
What about just storing the frame position directly and still removing `PlayPos`?

regulus79 replied:
I have removed `PlayPos`. Now each `Timeline` stores it's current frame offset and jumped state. I will see if I can get `jumped` to be passed in the `positionChanged` signal instead.

regulus79 replied:
I have removed the `m_jumped` member variable from `Timeline`. Now jumped events are handled with a `positionJumped` signal, and automatically emitted when the timeline ticks are forcefully set.

regulus79 commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -273,23 +270,24 @@ void AutomationEditor::keyPressEvent(QKeyEvent * ke )
 
 		case Qt::Key_Left:
-			if( ( m_timeLine->pos() -= 16 ) < 0 )
+			if(m_timeLine->model()->getTicks() - 16 > 0)
```
`Timeline` does not implement the `-=` operator

regulus79 commented on include/Song.h:
```diff
@@ -158,7 +129,7 @@ class LMMS_EXPORT Song : public TrackContainer
 	inline void setToTime(TimePos const & pos, PlayMode playMode)
 	{
 		m_elapsedMilliSeconds[static_cast<std::size_t>(playMode)] = pos.getTimeInMilliseconds(getTempo());
-		getPlayPos(playMode).setTicks(pos.getTicks());
+		getTimeline(playMode).setTicks(pos.getTicks());
```
`m_elapsedMilliseconds` is returned in `Song::getMillisends`, and afaik is only when the current time lcd is being updated. This could be removed and instead be calculated on the fly using the functions in `TimePos`, but if the tempo is changed during the song, this will not give the correct result.

`m_elapsedBars` and `m_elapsedTicks` do not appear to be used anywhere in the codebase, so I suppose they can be deleted?

regulus79 replied:
I have removed `m_elapsedMilliseconds`, `m_elapsedBars`, and `m_elapsedTicks`. The elapsed miliseconds used for the time lcd at the top are now calculated using the current frames. NOTE: This means that changing the tempo mid-song will cause the lcd to display the incorrect time.

regulus79 commented:
Partially following sakertooth's suggestion, I have removed `PlayPos.h` from the PR. Now each `Timeline` stores it's own `TimePos` along with a frame offset and `jumped` variable.

I have also removed `m_elapsedMilliseconds`, `m_elapsedBars`, and `m_elapsedTicks` from `Song`.

I could not find any code referencing `m_elapsedBars` or `m_elapsedTicks` in the entire codebase (outside of the code to update them), so removing those shouldn't cause any issues.

`m_elapsedMilliseconds` is used in displaying the current time LCD at the top. The current time can easily be calculated using the current frame, so removing this should also be fine in most cases, *however*, if you change the tempo mid-song, it will not display the correct time.

sakertooth also suggested getting rid of the `jumped` variable and instead passing it along with the `positionChanged` signal. I am still thinking of how that would be implemented, since `Song` would somehow have to store the `jumped` state when it receives the signal so that it can operate on it during the next buffer(?)

regulus79 commented:
I had to make some minor changes to fix the position line update its position when scrolling or zooming the piano roll/song editor.

Previously, it appears we were relying on the 60Hz QTimer to step in and update the position after scrolling or zooming. However, now that the timer is gone, we have to do it via signals.

If I've done everything right, there should be no visible changes from how it worked in 1.2.2.

messmerd commented on include/Song.h:
```diff
@@ -258,19 +226,12 @@ class LMMS_EXPORT Song : public TrackContainer
-	}
-	inline PlayPos & getPlayPos()
+
+	inline const TimePos & getPlayPos( PlayMode pm ) const
```
```suggestion
	const TimePos& getPlayPos(PlayMode pm) const
```

messmerd commented on include/Song.h:
```diff
@@ -258,19 +226,12 @@ class LMMS_EXPORT Song : public TrackContainer
+		return getTimeline(pm).getPlayPos();
 	}
-	inline const PlayPos & getPlayPos() const
+	inline const TimePos & getPlayPos() const
```
```suggestion
	const TimePos& getPlayPos() const
```

messmerd commented on include/Song.h:
```diff
@@ -147,7 +117,7 @@ class LMMS_EXPORT Song : public TrackContainer
 	inline int getMilliseconds(PlayMode playMode) const
 	{
-		return m_elapsedMilliSeconds[static_cast<std::size_t>(playMode)];
+		return 1000 * getFrames() / Engine::audioEngine()->outputSampleRate();
```
From what I understand, `m_elapsedMilliSeconds` and related variables are used to display the time elapsed at the current playhead position, though the way it's calculated is just a finicky estimation rather than an accurate value (which would require tracking tempo changes over time and calculating the elapsed time from that). Is that accurate?

Is there a good reason for removing `m_elapsedMilliSeconds` and friends from this PR, or should that wait for a future PR that properly fixes the elapsed time? I'm a little worried that this PR might just make it worse.

regulus79 replied:
> the way it's calculated is just a finicky estimation rather than an accurate value (which would require tracking tempo changes over time and calculating the elapsed time from that). Is that accurate?

Basically yes, at least in the current state of the PR. Before this PR, the elapsed time was calculated by incrementally adding on the "delta time" each period. This method is more or less immune to tempo changes/automation, since it doesn't depend on the global position of the playhead. (Well, at least for `m_elapsedMilliSeconds`. The bars/ticks appear to have been calculated from the playhead position)

However, if there is no tempo automation, it should be identical to just getting the current play head position and using that to calculate the elapsed time. Because of that, someone suggested that I remove it and just use that inline calculation instead.

The only place `Song::getMilliseconds` is used is in `TimeDisplayWidget`. If we do want the timer at the top toolbar to keep track of tempo changes, then I can add it back if you want. Oddly enough however, `m_elapsedTicks` and `m_elapsedBars` are not used anywhere in the entire codebase, so there should be no effect in removing them.

allejok96 reviewed (approved):
Apart from some minor things it looks good. I gave it a test run and found no problems.

Make sure to pull in the latest changes from master because there will be build errors in Pattern editor.

regulus79 commented on src/gui/editors/TimeLineWidget.cpp:
```diff
@@ -335,16 +326,16 @@ void TimeLineWidget::mouseMoveEvent( QMouseEvent* event )
 		case Action::MovePositionMarker:
-			m_pos.setTicks(timeAtCursor.getTicks());
+			m_timeline->setTicks(timeAtCursor.getTicks());
 			Engine::getSong()->setToTime(timeAtCursor, m_mode);
```
I just realized, trying to add the millisecond counter back in, that `setToTime` was used to update/reset the counter when the user moved the timeline playhead.

If we want, we could change it so that the millisecond lcd only shows the literal time spent playing the song and skips tracking any jumps in the playhead. Or I would have to make `TimeLineWidget` reach out directly to Song and update the time itself.

messmerd replied:
> If we want, we could change it so that the millisecond lcd only shows the literal time spent playing the song and skips tracking any jumps in the playhead.

That sounds like a regression.

> Or I would have to make TimeLineWidget reach out directly to Song and update the time itself.

Would it make sense to add `m_elapsedMilliSeconds` to `Timeline` instead of `Song`?

regulus79 replied:
> Would it make sense to add m_elapsedMilliSeconds to Timeline instead of Song?

That's a good idea.

I'm wondering if it can be nicely integrated with `Timeline`'s `setTicks` function. When incrementing the ticks, we want the offset amount and current tempo to be used to calculate the delta milliseconds. Based on the previous code, it seems that when dragging the playhead, it would also use the offset position and the current tempo to calculate the new milliseconds. However, when resetting the milliseconds when the playhead returns to the start, we don't want to use the current tempo, we want to directly set it to 0.

regulus79 replied:
I should ask: I know we want the timeline milliseconds to be updated based on the delta time as the buffers are being processed. This ensures the value shown on the lcd will be correct, no matter if the tempo is being automated or not.

However, what should the behavior be when the user drags the timeline playhead manually? Should the milliseconds be reset? Should they be calculated based on the current tempo and the new ticks? Should they be calculated based on the current tempo and the offset ticks (relative to the previous playhead position)?

allejok96 replied:
We want the correct time, but it's not so easy. So whatever solution gets us closer to that for most cases. I think current tempo+delta is good. That way it stays correct as long as you don't move past a tempo automation.

regulus79 replied:
Updating the millisecond counter based on the current tempo + delta would be nice, but looking at how 1.2.2 functioned, it seems it did it by the global playhead position and the current tempo whenever the user manually dragged the playhead (so dragging back to 0 would still show a time of 0). When actually playing the song however, it would use the delta and the current tempo to update the elapsed frames, so when playing from start to finish, it would show the correct time.

I have updated the elapsed millisecond handling to act in the same way.

regulus79 commented:
> What's the status of this now? Is it ready for review again?

I believe I left off still working/thinking of a good way to keep the millisecond timer working with tempo automations. I have a `git stash` somewhere of where I left off, but I've been super busy lately so I haven't touched it in a bit. I will try to prioritize working on it.

allejok96 commented on src/gui/editors/TimeLineWidget.cpp:
```diff
@@ -49,25 +49,22 @@ namespace
-	updateTimer->start( 1000 / 60 );  // 60 fps
 	connect( Engine::getSong(), SIGNAL(timeSignatureChanged(int,int)),
 					this, SLOT(update()));
+	connect(m_timeline, &Timeline::positionChanged, this, qOverload<>(&QWidget::update), Qt::QueuedConnection);
```
I did some crude testing, I mean very crude, and `Qt::DirectConnection` was 10x slower than a direct function call. `Qt::QueuedConnection` was 160x slower than a direct function call. Qt claims that not specifying the connection type will select the best one, while staying thread safe. So there's no need to force it to always take the slow path.

allejok96 commented:
> QTimer to constantly query the core and update the timeline widget

Please just make sure it doesn't do what it does currently - redrawing the entire BB editor at 60 FPS even when the song is paused :smiling_face_with_tear:

Veratil commented:
> > Reversing the direction. The UI should query the audio thread what the current time is. Also Qt probably shouldn't be in the audio thread.
> 
> Would you be in favor of the gui using something like a QTimer to constantly query the core and update the timeline widget?

It should either update with paintEvent or slower. My vote is paintEvent at this time.


> > QTimer to constantly query the core and update the timeline widget
> 
> Please just make sure it doesn't do what it does currently - redrawing the entire BB editor at 60 FPS even when the song is paused 🥲

With this PR, there's not much we can do about that; it redraws when paintEvent triggers.

regulus79 commented:
> It should either update with paintEvent or slower. My vote is paintEvent at this time.

I was under the impression that `paintEvent` is normally only called when the widget needs to be redrawn, for example when the user interacts with it or if `update()` is forcefully called, but that otherwise, widgets do not naturally update themselves on their own. Would a QTimer be needed to call `paintEvent` every howevermany milliseconds to update the timeline widget, or would it be connected to a signal which is emitted when the playhead has moved, or...?

messmerd commented on include/Timeline.h:
```diff
@@ -43,6 +45,42 @@ class Timeline : public QObject, public JournallingObject
+		emit positionJumped();
+	}
+
+	auto getTicks() const -> tick_t { return m_pos.getTicks(); }
```
I think it's fine to hold off on those then

allejok96 commented on src/core/Song.cpp:
```diff
@@ -668,20 +644,18 @@ void Song::stop()
-				setToTime(timeline.playStartPosition());
+				timeline.setTicks(timeline.playStartPosition().getTicks());
 
 				timeline.setPlayStartPosition(-1);
```
I know this is not part of the PR, but the negative value of `playStartPositoin` seems to only be used here... To I guess not send a `jumped` signal if stop is pressed multiple times? Idk. If that was the case you could simply check if the value is actually any different in `setTicks` and not emit a signal if the value didn't actually change.

allejok96 replied:
No, I understand now it's to not move the playhead back to where it started from if you press stop when it's already stopped. Personally I'd check `isPlaying` instead of using a negative value... but whatever

allejok96 commented on src/gui/editors/SongEditor.cpp:
```diff
@@ -96,13 +96,11 @@ SongEditor::SongEditor( Song * song ) :
-	connect(this, &TrackContainerView::positionChanged, m_timeLine, &TimeLineWidget::updatePosition);
-	connect( m_timeLine, SIGNAL( positionChanged( const lmms::TimePos& ) ),
-			this, SLOT( updatePosition( const lmms::TimePos& ) ) );
+	connect(this, &TrackContainerView::positionChanged, m_timeLine, qOverload<>(&QWidget::update));
```
Instead of connecting to this signal you could call `m_timeLine->update()` inside `SongEditor::scrolled()` - the only place I could find that emits this signal. Btw the signal will call `TrackContainerView::changePosition`

allejok96 commented on src/gui/editors/PatternEditor.cpp:
```diff
@@ -57,11 +57,11 @@ PatternEditor::PatternEditor(PatternStore* ps) :
+		m_currentPosition, this
 	);
-	connect(m_timeLine, &TimeLineWidget::positionChanged, this, &PatternEditor::updatePosition);
+	connect(this, &PatternEditor::positionChanged, m_timeLine, qOverload<>(&QWidget::update));
```
Btw `PatternEditor::updatePosition` will both update all clip views to display the blue moving thing, and also call `TrackContainerView::changePosition` that is used to switch between patterns - which is really bad design but totally unrelated to this PR... now where was I...

allejok96 commented on include/Timeline.h:
```diff
@@ -43,6 +45,42 @@ class Timeline : public QObject, public JournallingObject
+	// However, passing `jumped = false` will instead treat the change in ticks as a continuous increment
+	void setTicks(tick_t ticks, bool jumped = true)
+	{
+		if (jumped)
```
I think this function is better split in two:
- setTicks(ticks)
- increment/advanceTicks(delta)

since they differ both in second calculation and what signals they emit

allejok96 reviewed (changes requested):
Oh boy, I understand what you mean with having a hard time keeping track of connections. `positionChanged`, `updatePosition`, `scrolled` aaaaaaaAAaa

Reminder: all explicit `QueuedConnection` should be removed.

Apart for some minor stuff and general questions I found a bug during testing: when playing in piano roll with accompany, and moving the playhead in song editor to *before the clip* the playhead freezes in place in piano roll until it reaches the start of the clip.

regulus79 commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -1836,8 +1827,9 @@ void AutomationEditor::setTension()
+	const TimePos& t = m_timeLine->model()->pos();
 	if( ( Engine::getSong()->isPlaying() &&
 			Engine::getSong()->playMode() ==
 					Song::PlayMode::AutomationClip ) ||
```
Good point. Maybe in a future PR I could fix the automation editor timeline to perhaps sync itself to the parent editor timeline.

regulus79 commented:
> I found a bug during testing: when playing in piano roll with accompany, and moving the playhead in song editor to before the clip the playhead freezes in place in piano roll until it reaches the start of the clip.

This is also an issue in stable 1.2.2.

allejok96 commented on include/Timeline.h:
```diff
@@ -43,6 +45,41 @@ class Timeline : public QObject, public JournallingObject
 
+	auto pos() const -> const TimePos& { return m_pos; }
+
+	void setPlayPos(TimePos pos)
```
This function is reduntant (artifact from merging?), it doesn't update the second counter and the only usage of it passes a `tick_t`, not a `TimePos`.

regulus79 replied:
I originally had it so that it would line up with `Song::setPlayPos`, but now I realize that it is actually redundant, since `Song::setPlayPos` takes in ticks, so it doesn't matter, it can just call `Timeline::setTicks`. I've removed it in the most recent commit. :+1:

allejok96 commented on src/gui/editors/SongEditor.cpp:
Moving the playhead with the arrow keys is not possible when the song is paused, because it reads the position from `PlayMode::None` but writes to `PlayMode::Song`

```suggestion
		tick_t t = m_song->getTimeline(Song::PlayMode::Song).getTicks() - TimePos::ticksPerBar();
```

Idk that diff doesn't seem to work, but do the same from `Key_Right`

regulus79 commented on src/core/Song.cpp:
```diff
@@ -120,6 +117,15 @@ Song::Song() :
+	}
+
+	// Inform VST plugins and sample tracks if the user moved the play head
+	connect(this, &Song::playbackPositionJumped, [this](){ m_vstSyncController.setPlaybackJumped(true); });
```
As far as I know, Qt does not allow specifying the connection type when directly connecting to lambdas.

allejok96 commented on src/gui/editors/AutomationEditor.cpp:
```diff
@@ -131,13 +131,10 @@ AutomationEditor::AutomationEditor() :
-	connect(this, &AutomationEditor::positionChanged, m_timeLine, &TimeLineWidget::updatePosition);
-	connect( m_timeLine, SIGNAL( positionChanged( const lmms::TimePos& ) ),
-			this, SLOT( updatePosition( const lmms::TimePos& ) ) );
+	connect(m_timeLine->timeline(), &Timeline::positionChanged, this, &AutomationEditor::updatePosition, Qt::QueuedConnection);
```
```suggestion
	connect(m_timeLine->timeline(), &Timeline::positionChanged, this, &AutomationEditor::updatePosition);
```

allejok96 commented on src/gui/editors/PatternEditor.cpp:
```diff
@@ -57,11 +57,10 @@ PatternEditor::PatternEditor(PatternStore* ps) :
+		m_currentPosition, this
 	);
-	connect(m_timeLine, &TimeLineWidget::positionChanged, this, &PatternEditor::updatePosition);
+	connect(m_timeLine->timeline(), &Timeline::positionChanged, this, &PatternEditor::updatePosition, Qt::QueuedConnection);
```
```suggestion
	connect(m_timeLine->timeline(), &Timeline::positionChanged, this, &PatternEditor::updatePosition);
```

allejok96 commented on src/gui/editors/TimeLineWidget.cpp:
```diff
@@ -46,25 +46,21 @@ namespace
-	updateTimer->start( 1000 / 60 );  // 60 fps
 	connect( Engine::getSong(), SIGNAL(timeSignatureChanged(int,int)),
 					this, SLOT(update()));
+	connect(m_timeline, &Timeline::positionChanged, this, qOverload<>(&QWidget::update), Qt::QueuedConnection);
```
```suggestion
	connect(m_timeline, &Timeline::positionChanged, this, qOverload<>(&QWidget::update));
```
